File: src/example_cpp_node/src/publisher_node.cpp
/**
 * @file publisher_node.cpp
 * @brief Example ROS2 publisher node using custom messages in C++
 */

#include <chrono>
#include <memory>

#include "example_msgs/msg/example_message.hpp"
#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;

class ExamplePublisher : public rclcpp::Node {
 public:
  ExamplePublisher() : Node("example_publisher_cpp"), count_(0) {
    publisher_ = this->create_publisher<example_msgs::msg::ExampleMessage>(
        "example_topic_cpp", 10);
    timer_ = this->create_wall_timer(
        1000ms, std::bind(&ExamplePublisher::timer_callback, this));
    RCLCPP_INFO(this->get_logger(), "Example Publisher C++ Node started");
  }

 private:
  void timer_callback() {
    auto message = example_msgs::msg::ExampleMessage();
    message.name = "cpp_message_" + std::to_string(count_);
    message.value = count_;
    message.timestamp = this->now().seconds();
    message.is_active = true;

    RCLCPP_INFO(this->get_logger(), "Publishing: name='%s', value=%d, timestamp=%f",
                message.name.c_str(), message.value, message.timestamp);
    publisher_->publish(message);
    count_++;
  }

  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<example_msgs::msg::ExampleMessage>::SharedPtr publisher_;
  int count_;
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ExamplePublisher>());
  rclcpp::shutdown();
  return 0;
}

File: src/example_py_node/example_py_node/publisher_node.py
#!/usr/bin/env python3
"""Example ROS2 publisher node using custom messages."""

import time

import rclpy
from example_msgs.msg import ExampleMessage
from rclpy.node import Node


class ExamplePublisher(Node):
    """Publishes example messages at a fixed rate."""

    def __init__(self):
        super().__init__("example_publisher")
        self.publisher_ = self.create_publisher(ExampleMessage, "example_topic", 10)
        self.timer = self.create_timer(1.0, self.timer_callback)
        self.count = 0
        self.get_logger().info("Example Publisher Node started")

    def timer_callback(self):
        """Publish a message on each timer tick."""
        msg = ExampleMessage()
        msg.name = f"message_{self.count}"
        msg.value = self.count
        msg.timestamp = time.time()
        msg.is_active = True
        self.publisher_.publish(msg)
        self.get_logger().info(
            f'Publishing: name="{msg.name}", value={msg.value}, timestamp={msg.timestamp}'
        )
        self.count += 1


def main(args=None):
    """Main entry point for the publisher node."""
    rclpy.init(args=args)
    node = ExamplePublisher()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
